### PR TITLE
Issue #218 - Honoring context.xml when deploying (if present)

### PR DESCRIPTION
**Pull request by JustAnotherSoftwareDeveloper**
This is for issue #218 
First off let me state that I'm **BIG** on the Code Review as a mandatory step in every workflow that isn't a 100% throwaway project. Don't feel obligated to nitpick, but definitely open up about anything that doesn't sit right.

The workflow I decided on is such
1) If there is a context.xml present in the WAR file or webapp folder, vscode-tomcat will read in the path (again, if present)
2) vscode-tomcat will alter the appName to match the path, plus any modifications tomcat needs to put the application in the correct place when copying over the project to the webapps folder. (ie, path "/" goes to "ROOT"
3) If there is no context.xml, or path, then  vscode-tomcat will simpy default to the previous way of naming the application

Note:  I had to change the tslint config to allow "any" due to my use of parseXml, which is now public

I tested this on my folders and wars at work. All deploy correctly to the expected path. That being said, I've always been bad at smoke testing.

### CODE REVIEW

File: src/Tomcat/TomcatController.ts [hunk 1]
```diff
@@ -361,6 +364,39 @@ export class TomcatController {
         return path.extname(filePath).toLocaleLowerCase() === '.war';
     }
 
+    private async determineAppName(webappPath: string, server: TomcatServer): Promise<string> {
+        const defaultName: string = path.basename(webappPath, path.extname(webappPath));
+        let appName: string = defaultName;
+        let folderLocation: string;
+        if (this.isWarFile(webappPath)) {
+            folderLocation = path.join(this._tomcatModel.defaultStoragePath, defaultName);
+            if (fse.pathExistsSync(folderLocation)) {
```
Comment (SummerSun) [inline src/Tomcat/TomcatController.ts]: await fse.pathExists(folderLocation), awalys use async function call

File: tslint.json [hunk 2]
```diff
@@ -36,7 +36,7 @@
 			"match-default-export-name": true,
 			"mocha-avoid-only": true,
 			"mocha-no-side-effect-code": true,
-			"no-any": true,
+			"no-any": false,
```
Comment (SummerSun) [inline tslint.json]: just disable the no-any where you need it, in the line above or a block, but not disable all no-any checks

File: src/Tomcat/TomcatController.ts [hunk 3]
```diff
@@ -361,6 +364,39 @@ export class TomcatController {
...
+            }
+            fse.mkdirSync(folderLocation);
+            await Utility.executeCMD(this._outputChannel, server.getName(), 'jar', { cwd: folderLocation }, 'xvf', `${webappPath}`);
+        } else {
+            folderLocation = webappPath;
+        }
+        if (await fse.pathExists(path.join(folderLocation, 'META-INF', 'context.xml'))) {
+            const xml: string = fs.readFileSync(path.join(folderLocation, 'META-INF', 'context.xml'), 'utf8');
+            const jsonFromXml: any = await Utility.parseXml(xml);
+            if (jsonFromXml && jsonFromXml.Context && jsonFromXml.Context.$ && jsonFromXml.Context.$.path !== undefined) {
```
Comment (SummerSun) [inline src/Tomcat/TomcatController.ts]: lat undefined check not necessary

File: src/Tomcat/TomcatController.ts [hunk 4]
```diff
@@ -338,11 +338,14 @@ export class TomcatController {
         }
     }
 
-    private async deployWebapp(server: TomcatServer, webappPath: string): Promise<void> {
+    private async deployWebapp(
```
Comment (SummerSun) [inline src/Tomcat/TomcatController.ts]: not necessary change of the signature

**Comment by JustAnotherSoftwareDeveloper** [comment]
I went ahead and applied your suggestions. The one thing I didn't change is readFileSync(...) because the async version takes a callback, and does not return a Promise<string> . If that's an issue, let me know. 

As I said previously, I did some testing with WARs my company uses. However, tomcat is pretty complicated and there very well may be a config that messes this up.

File: src/Tomcat/TomcatController.ts [hunk 5]
```diff
@@ -361,6 +361,38 @@ export class TomcatController {
...
+        if (await fse.pathExists(path.join(folderLocation, 'META-INF', 'context.xml'))) {
+            const xml: string = fs.readFileSync(path.join(folderLocation, 'META-INF', 'context.xml'), 'utf8');
+            const jsonFromXml: any = await Utility.parseXml(xml);
+            if (jsonFromXml && jsonFromXml.Context && jsonFromXml.Context.$) {
+                const rawPath: string = jsonFromXml.Context.$.path;
+                appName = this.parseContextPathToFolderName(rawPath);
+            }
+        }
+        return appName;
+    }
```
Comment (SummerSun) [inline src/Tomcat/TomcatController.ts]: Enable no-any rule back after the function

**Comment by SummerSun** [comment]
Thank you for the changes and please make sure CI passes everytime you push more commits. Gald to see the functionality works in your uses. I will help with testing also after we merge it.

**Comment by x4nt** [comment]
@JustAnotherSoftwareDeveloper @SummerSun Is there any traction left on this PR? I would love to use this functionality, thank you!

**Comment by JustAnotherSoftwareDeveloper** [comment]
Okay full disclosure: I forgot all about this. I'll get my changes up with master and apply the commit changes sometime this week.

**Comment by SummerSun** [comment]
@JustAnotherSoftwareDeveloper Thanks for the contribution.

**Pull request merged**